### PR TITLE
Cranelift: do not check in generated ISLE code; regenerate on every compile.

### LINKED ISSUE #4066: ISLE: investigate always building from DSL (no checked-in source)
In our most recent Cranelift meeting, we discussed the question of whether we want to continue to rely on checked-in, pre-generated ISLE DSL compiler output, or else move back to a build strategy that always builds from the DSL.

The two main advantages of the current situation are:
- The build time of `cranelift-codegen` is low: there is only a little bit of `build.rs` work (the `meta`) crate, but we don't need to build the ISLE compiler every time.
- The generated source is present in the main tree and can be examined and debugged easily, to see why rules are or are not matching.

On the other hand, I believe that the following downsides are becoming more relevant:
- The workflow in general is very awkward. If we want to be friendly to newcomers who might want to tweak some instruction lowerings -- one of the main goals of ISLE, long-term -- then consider what understanding we assume:
  - We require the developer to build a mental model of the complex multi-step build with checked-in artifacts, which is unusual and unidiomatic in the Rust world;
  - We require a special Cargo flag to force a generated-source rebuild;
  - If there is ever more than one PR outstanding at a time to the ISLE DSL source, all but the first to merge will need to regenerate and rebase (most of us probably take understanding `git` for granted but this is a nontrivial speedbump for many);
  - The diffs that one sees on GitHub are cluttered with generated-source changes, which I have mostly learned to subconsciously ignore when reviewing but are potentially pretty confusing for onlookers.

The above factors are a mix of friction for newcomers (required understanding and complexity) and regular contributors (rebase pain). If the goal of the DSL is to make modifying the compiler easy, then the above factors I think are working against the goal of the DSL.

A quick survey of some other compilers and their build systems: of LLVM, SpiderMonkey and Golang, all of which have some meta-build steps, only Golang checks in generated source (and in the Go community this seems to be a bit more idiomatic).

So, to me, the question becomes how to support the needs above (build time, debugging) in a different way, if we can. I'd like to see if we can:
- Find a way to reduce the dependencies of `cranelift-isle` so it can be built in a few seconds at most. We use `miette` and `thiserror` (all for good reasons, to be clear!) and these pull in a bunch of (i) derive-macro stuff and (ii) fancy terminal stuff for `miette`. The latter can be removed when not in a "developer mode" but the derive-macros seem to go pretty deep. I don't think we should get rid of `miette` -- its errors are fantastic -- but could we refactor its use out to `islec`, propagating a more generic and minimal error-plus-span type out of `isle`? Ideally the crate tree used by `build.rs` should have just `cranelift-isle` and its one other dep `log`.
- Add a means to print "tracing" messages (off by default) in generated code, to ease debugging of why rules did or did not match.

Thoughts? cc @abrown @fitzgen

### PR DESCRIPTION
**Pull request by cfallin**
This PR fixes #4066: it modifies the Cranelift `build.rs` workflow to
invoke the ISLE DSL compiler on every compilation, rather than only
when the user specifies a special "rebuild ISLE" feature.

The main benefit of this change is that it vastly simplifies the mental
model required of developers, and removes a bunch of failure modes
we have tried to work around in other ways. There is now just one
"source of truth", the ISLE source itself, in the repository, and so there
is no need to understand a special "rebuild" step and how to handle
merge errors. There is no special process needed to develop the compiler
when modifying the DSL. And there is no "noise" in the git history produced
by constantly-regenerated files.

The two main downsides we discussed in #4066 are:
- Compile time could increase, by adding more to the "meta" step before the main build;
- It becomes less obvious where the source definitions are (everything becomes
  more "magic"), which makes exploration and debugging harder.

This PR addresses each of these concerns:

1. To maintain reasonable compile time, it includes work to cut down the
   dependencies of the `cranelift-isle` crate to *nothing* (only the Rust stdlib),
   in the default build. It does this by putting the error-reporting bits
   (`miette` crate) under an optional feature, and the logging (`log` crate) under
   a feature-controlled macro, and manually writing an `Error` impl rather than
   using `thiserror`. This completely avoids proc macros and the `syn` build slowness.

   The user can still get nice errors out of `miette`: this is enabled by specifying
   a Cargo feature `--features isle-errors`.

2. To allow the user to optionally inspect the generated source, which nominally
   lives in a hard-to-find path inside `target/` now, this PR adds a feature `isle-in-source-tree`
   that, as implied by the name, moves the target for ISLE generated source into
   the source tree, at `cranelift/codegen/isle_generated_source/`. It seems reasonable
   to do this when an explicit feature (opt-in) is specified because this is how ISLE regeneration
   currently works as well. To prevent surprises, if the feature is *not* specified, the
   build fails if this directory exists.

The compile-time measures as follows on my system: a clean build (`cargo clean` first
then `cargo build -p cranelift-codegen`) on a MacBook M1 Air (macOS/aarch64) gives:

Before:

    Finished dev [unoptimized + debuginfo] target(s) in 8.26s
    cargo build -p cranelift-codegen  22.29s user 2.05s system 291% cpu 8.348 total

After:

    Finished dev [unoptimized + debuginfo] target(s) in 9.96s
    cargo build -p cranelift-codegen  28.08s user 2.43s system 303% cpu 10.055 total

so about 1.7 seconds slower.

With a default configuration, an ISLE error looks like this during build:

```
  --- stderr
  Error building ISLE files: ISLE errors:

  parse error: Unexpected token Symbol("asdf")

  To see a more detailed error report, run:

      $ cargo check -p cranelift-codegen --features isle-errors
```

When specifying the feature, one instead gets

```
  --- stderr
  Error building ISLE files:
    × parse error: Unexpected token Symbol("asdf")
     ╭─[src/isa/aarch64/inst.isle:1:1]
   1 │ asdf
     · ┬
     · ╰── Unexpected token Symbol("asdf")
   2 │
     ╰────
```

### CODE REVIEW

File: cranelift/isle/isle/Cargo.toml [hunk 1]
```diff
@@ -9,9 +9,14 @@ repository = "https://github.com/bytecodealliance/wasmtime/tree/main/cranelift/i
...
+log = { version = "0.4", optional = true }
+miette = { version = "4.7.0", optional = true }
 
 [dev-dependencies]
 tempfile = "3"
+
+[features]
+default = []
+
+logging = []
```
Comment (abrown) [inline cranelift/isle/isle/Cargo.toml:21]:
```suggestion
logging = ["log"]
```

**Review by abrown — APPROVED** [review]
Makes sense to me; it will be a bit harder to remember to use these features if/when things go wrong but that's what documentation is for!

File: cranelift/isle/isle/Cargo.toml (hunk 1 above)
Comment (cfallin) [reply cranelift/isle/isle/Cargo.toml]: Fixed, thanks!

**Comment by cfallin** [comment]
Thanks! Updated and fixed the deterministic-meta test (previously was skipping ISLE build entirely, now does so again); will merge with your r+.

**Comment by bjorn3** [comment]
For me cranelift-isle takes 10s to build. The full `cargo check -p cranelift-codegen` takes 25s. `cargo build -p cranelift-codegen` takes 35s and `cargo build -p cranelift-codegen --release` takes 50s. So it is not ideal, but not the biggest problem either. I will see if there are some easy wins to have.

**Comment by fitzgen** [comment]
> With a default configuration, an ISLE error looks like this during build:
> 
> ```
>   --- stderr
>   Error building ISLE files: ISLE errors:
> 
>   parse error: Unexpected token Symbol("asdf")
> 
>   To see a more detailed error report, run:
> 
>       $ cargo check -p cranelift-codegen --features isle-errors
> ```

Any chance we could still have a source position in there when the `isle-errors` feature is not enabled?

**Comment by cfallin** [comment]
> Any chance we could still have a source position in there when the `isle-errors` feature is not enabled?

Sure! #4151

**Pull request merged**